docs/journey-engine-concepts.md

```diff
@@ -0,0 +1,138 @@
+# Journey Engine
+
+## Terms
+
+### Journey Engine
+
+The Journey Engine is what tells IPV Core Front what to show the user next. IPV Core Front calls the journey engine with `/journey/<journey-event>` requests to the internal API Gateway and the Journey Engine returns a Journey Response, that tells IPV Core front to show a page in IPV Core, redirect to a CRI, or a redirect to the originating OAuth client. 
+
+(The Journey Engine can also tell IPV Core Front to fire a new Journey Event back to the Journey Engine, though this is being phased out.)
+
+### Journey Engine Step Function
+
+This is the AWS Step Function that orchestrates the running of the lambdas that make up the Journey Engine.
+
+### Journey Engine State Machine
+
+The Journey Engine State Machine is a custom build implementation of an event driven state machine. This transitions the user from one state to another when Journey Events are received. It is configured using a Journey Map
+
+### Journey Map
+
+The Journey Map defines what sates the user can be in, the events that those states accept and what states those events transition to. It also specifies what should happen during those transition. The Journey Map is the configuration of the Journey Engine State Machine and consists of a collection of Journey States.
+
+### Nested Journey Map
+
+To aid in building up a complex Journey Map, a smaller collection of Journey States can be referenced. This is a Nested Journey Map.
+
+### Journey State
+
+A Journey Sate represents a point in the user's journey. The definition includes the events that can be accepted when the user is in this state and what happens when that event is received.
+
+### Journey Event
+
+A Journey Event is a named event that has occurred in the users journey. IPV Core Front is responsible for interpreting the action a user has taken and turning it into an event that can be sent to the Journey Engine. The event is represented as a string in requests to the internal API Gateway `/journey/<journey-event>`
+
+### Journey Response
```

Comment:
There's something else here too that overlaps quite a lot and I think can be the source of some confusion.

When we transition states in the state machine we define a "response" which is what the process-journey-step lambda returns. They are `JourneyStepResponse`s and you can see the types (page, journey, error, cri), defined  [here](https://github.com/alphagov/di-ipv-core-back/tree/main/lambdas/process-journey-step/src/main/java/uk/gov/di/ipv/core/processjourneystep/statemachine/responses). The journey responses from here are initially used by the step function to do some routing. OR (I think), returned to the frontend. The CRI response defined here is actually a journey response that is constructed slightly differently and then interpreted by the step function to run the build-oauth-request lambda. Confusing. The page and error responses get sent to the frontend.

We also define a different set of responses that have similar types and are derived from a [`BaseResponse`](https://github.com/alphagov/di-ipv-core-back/blob/main/lib/src/main/java/uk/gov/di/ipv/core/library/domain/BaseResponse.java). The types are JourneyResponse, ProvenUserIdentityDetails, CriResponse, ClientResponse. These tend to be the returns of lambdas themselves.

We should differentiate these things and make it obvious what they are.